storage/localstore/mode_set_test.go

```diff
@@ -63,44 +63,167 @@ func TestModeSetAccess(t *testing.T) {
+		{
+			name:            "set push sync, anonymous tag, with pinning",
+			anonymous:       true,
+			pin:             true,
+			mode:            chunk.ModeSetSyncPush,
+			runGcIndexTest:  true,
+			expErrPushIndex: leveldb.ErrNotFound,
+			expErrGCIndex:   nil,
+			expErrPinIndex:  nil,
+		},
+	} {
+		t.Run(mtc.name, func(t *testing.T) {
+			for _, tc := range multiChunkTestCases {
+				t.Run(tc.name, func(t *testing.T) {
+					db, cleanupFunc := newTestDB(t, &Options{Tags: chunk.NewTags()})
+					defer cleanupFunc()
+
+					tag, err := db.tags.Create(mtc.name, int64(tc.count), mtc.anonymous)
+					if err != nil {
+						t.Fatal(err)
+					}
+					if tag.Uid != 0 {
+						t.Fatal("expected mock tag uid")
+					}
 
-			chunks := generateTestRandomChunks(tc.count)
+					chunks := generateTestRandomChunks(tc.count)
 
-			wantTimestamp := time.Now().UTC().UnixNano()
-			defer setNow(func() (t int64) {
-				return wantTimestamp
-			})()
+					wantTimestamp := time.Now().UTC().UnixNano()
+					defer setNow(func() (t int64) {
+						return wantTimestamp
+					})()
 
-			_, err := db.Put(context.Background(), chunk.ModePutUpload, chunks...)
-			if err != nil {
-				t.Fatal(err)
-			}
+					_, err = db.Put(context.Background(), chunk.ModePutUpload, chunks...)
+					if err != nil {
+						t.Fatal(err)
+					}
 
-			err = db.Set(context.Background(), chunk.ModeSetSync, chunkAddresses(chunks)...)
-			if err != nil {
-				t.Fatal(err)
-			}
+					tag.Total = int64(tc.count)
+					tag.Stored = int64(tc.count)
 
-			binIDs := make(map[uint8]uint64)
+					err = db.Set(context.Background(), mtc.mode, chunkAddresses(chunks)...)
+					if err != nil {
+						t.Fatal(err)
+					}
+					tagCounterTest(t, tc.count, mtc.mode, tag)
+					if mtc.pin {
+						err = db.Set(context.Background(), chunk.ModeSetPin, chunkAddresses(chunks)...)
+						if err != nil {
+							t.Fatal(err)
+						}
+					}
 
-			for _, ch := range chunks {
-				po := db.po(ch.Address())
-				binIDs[po]++
+					binIDs := make(map[uint8]uint64)
 
-				newRetrieveIndexesTestWithAccess(db, ch, wantTimestamp, wantTimestamp)(t)
-				newPushIndexTest(db, ch, wantTimestamp, leveldb.ErrNotFound)(t)
-				newGCIndexTest(db, ch, wantTimestamp, wantTimestamp, binIDs[po])(t)
-			}
+					for _, ch := range chunks {
+						po := db.po(ch.Address())
+						binIDs[po]++
 
-			t.Run("gc index count", newItemsCountTest(db.gcIndex, tc.count))
+						newRetrieveIndexesTestWithAccess(db, ch, wantTimestamp, wantTimestamp)(t)
+						newPullIndexTest(db, ch, binIDs[po], nil)(t)
+						newPushIndexTest(db, ch, wantTimestamp, mtc.expErrPushIndex)(t)
+						newGCIndexTest(db, ch, wantTimestamp, wantTimestamp, binIDs[po], mtc.expErrGCIndex)(t)
+						newPinIndexTest(db, ch, mtc.expErrPinIndex)(t)
 
-			t.Run("gc size", newIndexGCSizeTest(db))
+						// if the upload is anonymous then we expect to see some values in the gc index
+						if mtc.anonymous {
+							t.Run("gc index count", newItemsCountTest(db.gcIndex, tc.count))
```

Comment:
This subtest prints a line in test output for every chunk which go up to the thousand. It is better not to call a subtest just a function.